heir_py/mlir_emitter.py

```diff
@@ -0,0 +1,199 @@
+                return self.emit_return(instr)
+        raise NotImplementedError("Unsupported instruction: " + str(instr))
+
+    def get_or_create_name(self, var):
+        name = var.name
+        if name in self.numba_names_to_ssa_var_names:
+            ssa_id = self.numba_names_to_ssa_var_names[name]
+        else:
+            ssa_id = self.temp_var_id
+            self.numba_names_to_ssa_var_names[name] = ssa_id
+            self.temp_var_id += 1
+        return f"%{ssa_id}"
+
+    def get_name(self, var):
+        assert var.name in self.numba_names_to_ssa_var_names
+        return self.get_or_create_name(var)
+
+    def forward_name(self, from_var, to_var):
+        to_name = self.numba_names_to_ssa_var_names[to_var.name]
+        self.numba_names_to_ssa_var_names[from_var.name] = to_name
+
+    def emit_assign(self, assign):
+        match assign.value:
+            case ir.Arg():
+                if assign.target.name != assign.value.name:
+                    raise ValueError(
+                        "MLIR has no vanilla assignment op? "
+                        "Do I need to keep a mapping from func arg names to SSA names?"
+                    )
+                self.numba_names_to_ssa_var_names[
+                    assign.target.name
+                ] = assign.value.name
+                return ""
+            case ir.Expr(op="binop"):
+                name = self.get_or_create_name(assign.target)
+                emitted_expr = self.emit_binop(assign.value)
+                # FIXME: replace i64 with inferred type
+                return f"{name} = {emitted_expr} : i64"
+            case ir.Expr(op="call"):
+                func = assign.value.func
+                # if assert fails, variable was undefined
+                assert func.name in self.globals_map
+                if self.globals_map[func.name] == "bool":
+                    # nothing to do, forward the name to the arg of bool()
+                    self.forward_name(
+                        from_var=assign.target, to_var=assign.value.args[0]
+                    )
+                    return ""
+                else:
+                    raise NotImplementedError("Unknown global " + func.name)
+            case ir.Expr(op="cast"):
+                # not sure what to do here. maybe will be needed for type conversions
+                # when interfacing with C
+                self.forward_name(from_var=assign.target, to_var=assign.value.value)
+                return ""
+            case ir.Const():
+                name = self.get_or_create_name(assign.target)
+                # FIXME: fix type (somehow the pretty printer on assign.value knows it's
+                # an int???)
+                return f"{name} = arith.constant {assign.value.value} : i64"
+            case ir.Global():
+                self.globals_map[assign.target.name] = assign.value.name
+                return ""
+        raise NotImplementedError()
+
+    def emit_expr(self, expr):
```

Comment:
Alternative approach: store all of the strings in a set or dict and use something like: 
`if expr.op in set:
    raise NotImplementedError() `